src/pow.rs

```diff
@@ -0,0 +1,1540 @@
+        impl_from_unsigned_type!($ty, u128);
+
+        impl_try_from_signed_type!($ty, i8);
+        impl_try_from_signed_type!($ty, i16);
+        impl_try_from_signed_type!($ty, i32);
+        impl_try_from_signed_type!($ty, i64);
+        impl_try_from_signed_type!($ty, i128);
+
+        impl fmt::Display for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+        }
+
+        impl fmt::LowerHex for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::LowerHex::fmt(&self.0, f) }
+        }
+
+        impl fmt::UpperHex for $ty {
+            #[inline]
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::UpperHex::fmt(&self.0, f) }
+        }
+
+        impl_div_mul!($ty, u8);
+        impl_div_mul!($ty, u16);
+        impl_div_mul!($ty, u32);
+        impl_div_mul!($ty, u64);
+
+        impl Shl<u32> for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn shl(self, shift: u32) -> Self { Self(self.0 << shift) }
+        }
+
+        impl Shr<u32> for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn shr(self, shift: u32) -> Self { Self(self.0 >> shift) }
+        }
+    };
+}
+
+/// A 256 bit integer representing work.
+///
+/// Work is a measure of how difficult it is to find a hash below a given [`Target`].
+///
+/// ref: <https://en.bitcoin.it/wiki/Work>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Work;
+/// let _ = Work::from(0xdeadbeef_u32);
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Work(Uint256);
+
+impl Work {
+    /// Lowest possible work value for Mainnet. See comment on [`Params::pow_limit`] for more info.
+    pub const MAINNET_MIN: Work =
+        Work(Uint256([0x0000_0000_ffff_0000_0000_0000_0000_0000_u128, 0]));
+
+    /// Lowest possible work value for Testnet. See comment on [`Params::pow_limit`] for more info.
+    pub const TESTNET_MIN: Work =
+        Work(Uint256([0x0000_0000_ffff_0000_0000_0000_0000_0000_u128, 0]));
+
+    /// Lowest possible work value for Signet. See comment on [`Params::pow_limit`] for more info.
+    pub const SIGNET_MIN: Work = Work(Uint256([0x0000_0377_ae00_0000_0000_0000_0000_0000_u128, 0]));
+
+    /// Lowest possible work value for Regtest. See comment on [`Params::pow_limit`] for more info.
+    pub const REGTEST_MIN: Work =
+        Work(Uint256([0x7fff_ff00_0000_0000_0000_0000_0000_0000_u128, 0]));
+
+    /// Converts this [`Work`] to [`Target`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Work;
+    /// let w = Work::from(0xdeadbeef_u32);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(self) -> Target { Target(self.0.inverse()) }
```

Comment:
I gave up on const functions because Rust 1.41 is really restrictive e.g. no `if` statements, no `for` statements, and even for functions that can be const I figured making them so really restricts how we can modify the functions internally. Thoughts?

Reply:
Since we have not really started using const functions throughout the codebase it feels like something that can be done as a single block of work (PRs or epic) as opposed to now during this PR?

Reply:
> making them so really restricts how we can modify the functions internally. Thoughts?

What it mainly restricts is doing syscalls, which we will never do in integer types. There are other temporary restrictions depending on Rust version, the most painful being lack of helpful APIs.

As you may have noticed I mainly wanted constructors on the grounds that it allows usage of the type in const contexts. Unfortunately it spilled due to some constructors using math.

While I do agree with the general idea of adding `const` later, I'd prefer to have reasonably simple constructors `const` from the beginning to avoid the overhead of several PRs. So for this PR, I'd like `from_le_bytes` and `from_be_bytes` to be `const`. It's enough to use the types in const context and people can calculate manually and feed the bytes if needed. While writing the cont code is a bit annoying, I did it for you, so you just need to copy it and review. :)

Reply:
I just saw this comment, will try to work this in to next version.